// ufs-mobile-platform/ios/build/UFSLibrary.framework/JSHeaders/Components/Visual/Input/Value/TextInput/Bridge/UFSTextInput.js
var __assign = (this && this.__assign) || Object.assign || function(t) {
    for (var s, i = 1, n = arguments.length; i < n; i++) {
        s = arguments[i];
        for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
            t[p] = s[p];
    }
    return t;
};
import React, { Component } from 'react';
import ReactNative, { LayoutAnimation, UIManager, requireNativeComponent } from 'react-native';
import { renderHintIcons } from '../../../../HintIcon/HintIcon/Bridge/UFSHintIcon';
import { renderInputLeftContainer } from '../../Container/InputLeftContainer/Bridge/UFSInputLeftContainer';
import { renderInputRightContainer } from '../../Container/InputRightContainer/Bridge/UFSInputRightContainer';
/**
 * Компонент предназначен для ввода текстовых или числовых данных.
 * ##### Размещение
 * При размещении в платформенных контейнерах(`SplitPanel`, `Popover`) будет
 * работать функция автоматического скролла при появлении и скрытии клавиатуры.
 *
 * ##### Методы
 * 1. dismissKeyboard()
 *
 * <br/>
 *
 * Принудительно скрывает клавиатуру.
 * @category Visual/Input
 */
export default class UFSTextInputManager extends Component {
    constructor() {
        super(...arguments);
        this.forceUpdateValue = (value) => {
            this.instance.setNativeProps({ value });
        };
        this.onChange = (event) => {
            this.oldValue = this.props.value;
            this.props.onChange && this.props.onChange(event.nativeEvent.value);
            this.forceUpdate();
        };
        /**
         * Принудительно скрывает клавиатуру.
         */
        this.dismissKeyboard = () => {
            UIManager.dispatchViewManagerCommand(ReactNative.findNodeHandle(this), UIManager.UFSTextInputView.Commands.dismissKeyboard, []);
        };
        this.onComponentRef = (component) => {
            this.instance = component;
        };
    }
    componentWillUpdate(nextProps, nextState) {
        if (nextProps.hasError !== this.props.hasError || nextProps.hasWarning !== this.props.hasWarning) {
            LayoutAnimation.easeInEaseOut();
        }
    }
    componentDidUpdate() {
        let value = this.props.value;
        if (this.oldValue == value) {
            this.forceUpdateValue(value);
        }
    }
    render() {
        return (React.createElement(UFSTextInput, __assign({ref: this.onComponentRef}, this.props, {onChange: this.onChange}), 
            renderInputLeftContainer(this.props.children), 
            renderInputRightContainer(this.props.children), 
            renderHintIcons(this.props.children, 'TextInput', 1)));
    }
}
export const UFSKeyboardType = UIManager.UFSTextInputView.Constants.KeyboardType;
export const UFSReturnKeyType = UIManager.UFSTextInputView.Constants.ReturnKeyType;
const UFSTextInput = requireNativeComponent('UFSTextInputView');
