// ufs-mobile-platform/ios/build/UFSLibrary.framework/JSHeaders/Components/Visual/Text/UFSPresets.js
import { StyleSheet } from 'react-native';
import { colors } from '../../Common';
// Text styles
export const textStyles = StyleSheet.create({
    display2: {
        fontFamily: "PTSans-Bold",
        fontSize: 28,
        lineHeight: 34.0,
        color: colors.grey900
    },
    display1: {
        fontFamily: "PTSans-Narrow",
        fontSize: 24,
        lineHeight: 34.0,
        color: colors.grey900
    },
    title: {
        fontFamily: "PTSans-Bold",
        fontSize: 20,
        lineHeight: 26.0,
        color: colors.grey900
    },
    heading: {
        fontFamily: "PTSans-Bold",
        fontSize: 16,
        lineHeight: 24.0,
        color: colors.grey900
    },
    paragraph2: {
        fontFamily: "PTSans-Bold",
        fontSize: 16,
        lineHeight: 24.0,
        color: colors.grey900
    },
    body2: {
        fontFamily: "PTSans-Bold",
        fontSize: 16,
        lineHeight: 20.0,
        color: colors.grey900
    },
    button2: {
        fontFamily: "PTSans-Bold",
        fontSize: 16,
        lineHeight: 20.0
    },
    body1: {
        fontFamily: "PTSans-Regular",
        fontSize: 16,
        lineHeight: 20.0,
        color: colors.grey900
    },
    paragraph1: {
        fontFamily: "PTSans-Regular",
        fontSize: 16,
        lineHeight: 24.0,
        color: colors.grey900
    },
    button1: {
        fontFamily: "PTSans-Regular",
        fontSize: 16,
        lineHeight: 20.0,
        color: colors.blue700
    },
    subheading: {
        fontFamily: "PTSans-Bold",
        fontSize: 14,
        lineHeight: 18.0,
        color: colors.grey900
    },
    caption: {
        fontFamily: "PTSans-Regular",
        fontSize: 12,
        lineHeight: 14.0,
        color: colors.grey500
    }
});
